sopel/test_tools.py

```diff
@@ -259,3 +272,257 @@ def run_example_tests(filename, tb='native', multithread=False, verbose=False):
+
+    This will instanciate a :class:`~sopel.bot.Sopel` object, replace its
+    backend with a :class:`~MockIRCBackend`, and then autoload plugins. This
+    will automatically load the ``coretasks`` plugins, and every other plugins
+    from ``autoloads``::
+
+        bot = botfactory(settings, ['emoticons', 'find', 'remind'])
+    """
+    autoloads = set(autoloads or []) | {'coretasks'}
+
+    mockbot = Sopel(settings)
+    mockbot.backend = MockIRCBackend(mockbot)
+
+    usable_plugins = sopel.plugins.get_usable_plugins(settings)
+    for name in autoloads:
+        plugin = usable_plugins[name][0]
+        plugin.load()
+        plugin.register(mockbot)
+
+    return mockbot
+
+
+class IRCFactory(object):
+    """Factory to send IRC messages from a fake server to a test bot.
+
+    :param bot: test bot instance to send messages to
+    :type bot: :class:`sopel.bot.Sopel`
+
+    This factory helps developpers when they want to simulate an IRC server
+    sending messages to the bot.
+
+    You can generate a test bot instance with the :func:`botfactory` function::
+
+        >>> from sopel import test_tools
+        >>> bot = test_tools.botfactory(settings, ['emoticons'])
+        >>> irc = test_tools.IRCFactory(bot)
+        >>> irc.say(test_tools.UserFactory('TestSubject', '#test', '.shrug'))
+        >>> bot.backend.message_sent
+        [b'PRIVMSG #test :¯\\_(ツ)_/¯']
+
+    .. seealso::
+
+        This factory is best used with all the testing tools available:
+
+        * :func:`botfactory` to create a test instance of Sopel
+        * :class:`UserFactory` to fake IRC user who will join channels, send
+          messages, use commands, trigger URL callback, etc.
+        * :class:`MockIRCBackend` which doesn't need a connection to any real
+          server and will store every messages sent by the bot in a convenient
+          place
+        * :func:`rawlist` to generate an expected output and make assertions
+
+    """
+    def __init__(self, bot):
+        self.bot = bot
+
+    @property
+    def chanserv(self):
+        """ChanServ's message prefix."""
+        return 'ChanServ!ChanServ@services.'
+
+    def channel_joined(self, channel, users=None):
+        """Send events as if the bot just joined a channel.
+
+        :param str channel: channel to send message for
+        :param list users: list (or tuple) of nicknames that will be present
+                           in the ``RPL_NAMEREPLY`` event
+
+        This will send 2 messages to the bot:
+
+        * a ``RPL_NAMEREPLY`` event (353), giving information about ``users``
+          present in ``channel``
+        * a ``RPL_ENDOFNAMES`` event (366) for completion
+
+        Use this to emulate when the bot joins to a channel, and the server
+        replies with the list of connected users::
+
+            factory.channel_joined('#test', ['Owner', '@ChanServ'])
+
+        in this example, the bot will know that there are 2 other users present
```

Comment:
Start of a sentence.

```suggestion
        In this example, the bot will know that there are 2 other users present
```